File: 064. Minimum Path Sum.scala
/*
Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right which minimizes the sum of all numbers along its path.

Note: You can only move either down or right at any point in time.

Seen this question in a real interview before?   Yes  
*/

def minPathSum(mark: Array[Array[Int]]): Int = {
    if(mark==null || mark.length==0 || mark(0).length==0) return 0;
    var m = mark.length;
    var n = mark(0).length;
    

    for(i<- 1 to m-1){
        mark(i)(0) += mark(i-1)(0)
    }
        
    for(j<- 1 to n-1){
        mark(0)(j) += mark(0)(j-1)
    }
    for(i<-1 to m-1){
        for(j<-1 to n-1){
            mark(i)(j) += math.min(mark(i)(j-1),mark(i-1)(j));
        } 
    }
    mark(m-1)(n-1);
}

File: 064. Minimum Path Sum.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m,n = len(grid),len(grid[0])
        matrix = [[0]*n for i in range(0,m)]
        matrix[0][0] = grid[0][0]
        
        for i in range(1,m):
            matrix[i][0] = matrix[i-1][0] + grid[i][0]
        for j in range(1,n):
            matrix[0][j] = matrix[0][j-1] + grid[0][j]
            
        for i in range(1,m):
            for j in range(1,n):
                matrix[i][j] = grid[i][j] + min(matrix[i-1][j],matrix[i][j-1])
        return matrix[m-1][n-1]
